widgets/filters/src/FilterActions.ts

```diff
@@ -0,0 +1,11 @@
+export default class FilterActions {
+    toolbox: Stage.Types.Toolbox;
+
+    constructor(toolbox: Stage.Types.Toolbox) {
+        this.toolbox = toolbox;
+    }
+
+    doList(params: unknown) {
```

Comment:
Do we know the type of params and the response type?

If yes, supplying it here would be great for readability and discoverability :slightly_smiling_face: Not having to jump to the devtools/docs/explore uses of the function is miles worse than having the type right here :slightly_smiling_face: 

```suggestion
    doList(params: ???): Promise<Stage.Types.PaginatedResonse<Filter> {
```

Reply:
My understanding is that here `params` is a black box that we are just passing further, without caring what it actually is. So, its type should be determined only by what `doGet` accepts - but it is not yet in TS so it accepts `any`.

Question: if `doGet` specified its parameters types then would it be necessary to specify type for `params`? Would type inference work here?

Regarding return type - I just added it.

Reply:
> My understanding is that here `params` is a black box that we are just passing further, without caring what it actually is. So, its type should be determined only by what `doGet` accepts - but it is not yet in TS so it accepts `any`.

The `any` type (could have been `unknown` though) in `doGet` is because `doGet` is very general. It should accept any URL, and thus, any parameters.

Here we have a specialized function - one that fetches the filters from a very specific URL. That API endpoint also has a known set of parameters that it accepts. Knowing them, we can do better than `any` or `unknown` :slightly_smiling_face: 

> Question: if `doGet` specified its parameters types then would it be necessary to specify type for `params`? Would type inference work here?

Unfortunately, not out of the box. You would have to import the type that `doGet` uses for parameter, and then use it here, or, do some magic with the `Parameters` type:

```ts
type Manager = ReturnType<Stage.Types.Toolbox['getManager']>;
type Params = Parameters<Manager['doGet']>[1];
```

I would recommend the first approach, though, if `doGet`'s parameters are ever strongly typed :smile:

Reply:
> The any type (could have been unknown though) in doGet is because doGet is very general. It should accept any URL, and thus, any parameters.

Not really. `params` are just query parameters, so the expected type is an object keyed by string. It's just not typed yet. If it was, the type for `params` would be naturally enforced.

> Here we have a specialized function - one that fetches the filters from a very specific URL. That API endpoint also has a known set of parameters that it accepts. Knowing them, we can do better than any or unknown 🙂

What for if nobody would use parameters defined by this type, at least not at this point?

But generally, looking at how the REST API works, I suspect it should be possible to create a general type describing query params - there are some common parts (pagination, sorting, searching, including specific properties) while the endpoint specific part is based on the response item type. So I imagine a parametrized type, like this: `QueryParams<Filter>`. However, since I don't need such type at the moment I'd rather not spend time on it now.

Reply:
> Not really. `params` are just query parameters, so the expected type is an object keyed by string. It's just not typed yet. If it was, the type for `params` would be naturally enforced.

Yeah, I guess `Record<string, any>` sounds reasonable there

> So I imagine a parametrized type, like this: QueryParams<Filter>

Makes sense

> However, since I don't need such type at the moment I'd rather not spend time on it now.

Sounds good. If the params accepted by the endpoint are the same as any other endpoint (essentially the `gridParams`), that's fine :+1: